keps/104-StartupPolicy/README.md

```diff
@@ -0,0 +1,437 @@
+possible to collect this information before implementation begins, in order to
+avoid requiring implementors to split their attention between writing release
+notes and implementing the feature itself. KEP editors and SIG Docs
+should help to ensure that the tone and content of the `Summary` section is
+useful for a wide audience.
+
+A good summary is probably at least a paragraph in length.
+
+Both in this section and below, follow the guidelines of the [documentation
+style guide]. In particular, wrap lines to a reasonable length, to make it
+easier for reviewers to cite specific portions, and to minimize diff churn on
+updates.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+
+
+
+## Motivation
+
+High Performance Computing / Machine Learning users usually want some control over how jobs are being created.  A Startup policy would allow a user to specify which event a job should have before starting other jobs in a jobset.  
+
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+### Goals
+- Add a startup policy for JobSet
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+### Non-Goals
+
+- Complicated workflow specifications
+  - This means that we will support simple dependecies rather than processing a full dag.
+  - Projects like argo-workflows are a better use case for this.
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### User Stories (Optional)
+
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+As a parallel framework user, I have one ReplicatedJob that functions as the driver and a series of worker ReplicatedJobs.  I want the driver to be ready before I start the worker replicated jobs.
+
+```yaml
+apiVersion: jobset.x-k8s.io/v1alpha2
+kind: JobSet
+metadata:
+  name: driver-ready-worker-start
+spec:
+  startupPolicy:
```

Comment:
Hmm it feels awkward to have a startupPolicy at the jobSet level that defines a target replicated job to wait for a certain status, before starting x/y/z replicated job(s)... the semantics are unclear, does it apply to all replicatedJobs which are not the target of the policy? It also doesn't seem to have a way to have certain replicated jobs wait on a certain replicated job, but not others. This flexibility might be nice for future use cases. 

What do you think?

Reply:
I was going to push thoughts on this.  

If StartupPolicy is specified, I would suspend all jobs.  And then unsuspend the ones that match the target.  And keep doing that as they are listed.

Reply:
Part of the issue with partial application is that I think the API gets pretty complicated as we operator on ReplicatedJobs rather than Jobs.  So we can have partially completiion of the replicatedJobs with this API.  

I think it would be nice to have a partial readiness at the job level but I tried to bring that up to #wg-batch but I think there was some light pushback as we could utilize the `ready` condition at the Job level.  

I notice that `ReplicatedJobStatus` gives you the overal status of the ReplicatedJob though.  I wasn't sure if this is a big enough change to fix that or not.  

I figured we can tackle a lot of the usecases specified in Stories without having to change ReplicatedJobStatus too much.

Reply:
I do think if it's at this level we would want to be able to target specific jobs. Arguably if a ReplicatedJob in a JobSet isn't linked to the other ones in a meaningful way, why would it need to be in the Jobset? 

It would make things simpler if at least the readiness was at the level of the Job. There are likely other use cases for that beyond Jobset.

Reply:
@ahg-g and @danielvegamyhre called that out originally in the original jobset doc.  I created https://github.com/kubernetes/kubernetes/issues/117758 for that reason.  

But there wasn't a large drive for that.  I think if we want something at the level of the job, we should bring that up as a use case for the job object.  

Right now, I went ahead with status ready field of Jobs.  This counts the number of pods that are ready and bubbles that information up to the job status.